Memory leak fix: pointers in interpolator and horiz_interp

laurenchilutti opened the pull request:
**Description**
In interpolator and horiz_interp, some pointers are being allocated.  This PR changes the type from pointer to allocatable.  Any use of `if associated()` has now been changed to `if allocated()` to accompany this change.  There are also instances in the code where we were declaring, for example:
```
interface assignment(=)
  module procedure horiz_interp_type_eq
end interface
```
And in that subroutine horiz_interp_type_eq, for example:
```
horiz_interp_out%faci => horiz_interp_in%faci
```

I have changed these pointer assignment operators to an equals, with the  foresight that this may cause issues.  With the old functionality, anytime `horiz_interp_out` is modified, the value of `horiz_interp_in` is also modified.  With my changes this will not be the case. I tested with the land model and did not run into any issues with this but we will test all models with alpha2 testing.

Fixes #1099 

**How Has This Been Tested?**
Compiled with GNU and Intel, ran with the land model and answers reproduce

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

laurenchilutti commented:
@rem1776 Could @ganganoaa be added as a reviewer for this PR? Thanks

thomas-robinson commented on horiz_interp/horiz_interp_type.F90:
```diff
@@ -189,35 +189,35 @@ subroutine horiz_interp_type_eq(horiz_interp_out, horiz_interp_in)
       call mpp_error(FATAL,'horiz_interp_type_eq: horiz_interp_type variable on right hand side is unassigned')
     endif
 
-    horiz_interp_out%faci            => horiz_interp_in%faci
```
The functionality of this routine is changing.  Can you add doxygen comments to this clearly stating that this routine creates a *copy* of the object.  Also doxygen comments for the variables.

laurenchilutti replied:
Let me know if these changes are sufficient

thomas-robinson commented on horiz_interp/horiz_interp_type.F90:
```diff
@@ -189,35 +189,35 @@ subroutine horiz_interp_type_eq(horiz_interp_out, horiz_interp_in)
+    horiz_interp_out%i_lon           =  horiz_interp_in%i_lon
+    horiz_interp_out%j_lat           =  horiz_interp_in%j_lat
+    horiz_interp_out%src_dist        =  horiz_interp_in%src_dist
     horiz_interp_out%found_neighbors => horiz_interp_in%found_neighbors
```
Did you miss this pointer?

laurenchilutti replied:
I couln't find any use of this pointer in FMS.  Specifically if you grep the repository, you only get the following lines:
https://github.com/NOAA-GFDL/FMS/blob/203c8bf464ff26fe0fe39b1451caedd026bbce55/horiz_interp/horiz_interp_type.F90#L72
https://github.com/NOAA-GFDL/FMS/blob/203c8bf464ff26fe0fe39b1451caedd026bbce55/horiz_interp/horiz_interp_type.F90#L203

So, since it wasn't a pointer being allocated I did not touch that variable at all.

laurenchilutti replied:
Does my reasoning make sense?

thomas-robinson replied:
It seems odd because the routine is creating a copy except for this one variable.

rem1776 replied:
I'd bet its just unused in general, you don't get warnings for unused fields in derived types so it could have never been used and no one noticed. I think we can probably remove it unless we can find some code that actually uses it.

ganganoaa replied:
If we remove this field assignment, it will violate principle of object-to-object copy but if it is never used, we can remove. We can add something like this:

if (associated(in%var)) 
    if (associated(out%var)) deallocate(out%var)
    allocate(out%var(size(in%var))
    out%var = in%var
else
    out%var => NULL()
end if

laurenchilutti replied:
@thomas-robinson @ganganoaa @rem1776 i have not done anything to found neighbors yet.  the options as I see it are:
- removing it seems like something that would require a lot of testing since its possible some model somewhere is using it.  
- changing it from pointer to allocatable change this pointer assignment to equals
- leave it as is
I don't feel like I have enough familiarity with interpolator code to make this decision.  How should I proceed?   Who can I consult on what to do?

thomas-robinson replied:
do this:
```fortran
if (allocated(horiz_interp_in%found_neighbors))horiz_interp_out%found_neighbors = horiz_interp_in%found_neighbors
```
and make `found_neighbors` allocatable instead of a pointer.  This is a `=` routine, so it should be making a copy, not providing pointers.  

This is a modernization PR, not a clean up.  No need to remove it here.  We should note it for potential clean up in a future release.

ganganoaa commented:
Yes. That was missed.

horiz_interp_out%found_neighbors => horiz_interp_in%found_neighbors


Ganga


On 1/3/23 09:15, Tom Robinson wrote:
>
> ***@***.**** requested changes on this pull request.
>
> ------------------------------------------------------------------------
>
> In horiz_interp/horiz_interp_type.F90 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1060620004>:
>
> > @@ -189,35 +189,35 @@ subroutine horiz_interp_type_eq(horiz_interp_out, horiz_interp_in)
>         call mpp_error(FATAL,'horiz_interp_type_eq: horiz_interp_type variable on right hand side is unassigned')
>       endif
>   
> -    horiz_interp_out%faci            => horiz_interp_in%faci
>
> The functionality of this routine is changing. Can you add doxygen 
> comments to this clearly stating that this routine creates a /copy/ of 
> the object. Also doxygen comments for the variables.
>
> ------------------------------------------------------------------------
>
> In horiz_interp/horiz_interp_type.F90 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1060620837>:
>
> > -    horiz_interp_out%wti             => horiz_interp_in%wti
> -    horiz_interp_out%wtj             => horiz_interp_in%wtj
> -    horiz_interp_out%i_lon           => horiz_interp_in%i_lon
> -    horiz_interp_out%j_lat           => horiz_interp_in%j_lat
> -    horiz_interp_out%src_dist        => horiz_interp_in%src_dist
> +    horiz_interp_out%faci            =  horiz_interp_in%faci
> +    horiz_interp_out%facj            =  horiz_interp_in%facj
> +    horiz_interp_out%ilon            =  horiz_interp_in%ilon
> +    horiz_interp_out%jlat            =  horiz_interp_in%jlat
> +    horiz_interp_out%area_src        =  horiz_interp_in%area_src
> +    horiz_interp_out%area_dst        =  horiz_interp_in%area_dst
> +    horiz_interp_out%wti             =  horiz_interp_in%wti
> +    horiz_interp_out%wtj             =  horiz_interp_in%wtj
> +    horiz_interp_out%i_lon           =  horiz_interp_in%i_lon
> +    horiz_interp_out%j_lat           =  horiz_interp_in%j_lat
> +    horiz_interp_out%src_dist        =  horiz_interp_in%src_dist
>       horiz_interp_out%found_neighbors => horiz_interp_in%found_neighbors
>
> Did you miss this pointer?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#pullrequestreview-1234603998>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/A43PSQH5HBRREIEUXKGZHJ3WQQYA5ANCNFSM6AAAAAATPWEFNQ>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

ganganoaa commented:
I think it is safe since it is not allocated any where.


On 1/4/23 10:26, laurenchilutti wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In horiz_interp/horiz_interp_type.F90 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1061600010>:
>
> > @@ -189,35 +189,35 @@ subroutine horiz_interp_type_eq(horiz_interp_out, horiz_interp_in)
>         call mpp_error(FATAL,'horiz_interp_type_eq: horiz_interp_type variable on right hand side is unassigned')
>       endif
>   
> -    horiz_interp_out%faci            => horiz_interp_in%faci
>
> Let me know if these changes are sufficient
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1061600010>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/A43PSQDBUJJA65UCFG3TCMLWQWJBXANCNFSM6AAAAAATPWEFNQ>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

ganganoaa commented on horiz_interp/horiz_interp_type.F90:
```diff
@@ -181,43 +181,44 @@ subroutine stats ( dat, low, high, avg, miss, missing_value, mask )
+    horiz_interp_out%wtj             =  horiz_interp_in%wtj
+    horiz_interp_out%i_lon           =  horiz_interp_in%i_lon
+    horiz_interp_out%j_lat           =  horiz_interp_in%j_lat
+    horiz_interp_out%src_dist        =  horiz_interp_in%src_dist
```
This and all the above pointers including one below are not allocated but their assignments are different compared to `found_neighbors' case.

thomas-robinson replied:
Where is that done?

laurenchilutti replied:
They are allocated in other horiz_interp files:

e.g.
https://github.com/laurenchilutti/FMS/blob/42e061f96343dd8e2df0d0338a372dc8c37b61a9/horiz_interp/horiz_interp_conserve.F90#L168-L171

https://github.com/laurenchilutti/FMS/blob/42e061f96343dd8e2df0d0338a372dc8c37b61a9/horiz_interp/horiz_interp_bilinear.F90#L288-L291

https://github.com/laurenchilutti/FMS/blob/42e061f96343dd8e2df0d0338a372dc8c37b61a9/horiz_interp/horiz_interp_spherical.F90#L217-L220

laurenchilutti replied:
But, found_neighbors is used and allocated nowhere in FMS that I can find

ganganoaa replied:
It is done here:
interface assignment(=)
  module procedure horiz_interp_type_eq
end interface

thomas-robinson commented on horiz_interp/horiz_interp_type.F90:
```diff
@@ -181,43 +181,44 @@ subroutine stats ( dat, low, high, avg, miss, missing_value, mask )
-    horiz_interp_out%j_lat           => horiz_interp_in%j_lat
-    horiz_interp_out%src_dist        => horiz_interp_in%src_dist
+    horiz_interp_out%faci            =  horiz_interp_in%faci
+    horiz_interp_out%facj            =  horiz_interp_in%facj
+    horiz_interp_out%ilon            =  horiz_interp_in%ilon
+    horiz_interp_out%jlat            =  horiz_interp_in%jlat
+    horiz_interp_out%area_src        =  horiz_interp_in%area_src
+    horiz_interp_out%area_dst        =  horiz_interp_in%area_dst
+    horiz_interp_out%wti             =  horiz_interp_in%wti
+    horiz_interp_out%wtj             =  horiz_interp_in%wtj
+    horiz_interp_out%i_lon           =  horiz_interp_in%i_lon
+    horiz_interp_out%j_lat           =  horiz_interp_in%j_lat
+    horiz_interp_out%src_dist        =  horiz_interp_in%src_dist
```
I think you need to check if the variable is allocated before you set it.
```fortran
    if (allocated(horiz_interp_in%wtj)) horiz_interp_out%wtj             =  horiz_interp_in%wtj
```

laurenchilutti replied:
I see what you are saying - do you think it would also be necessary to check if a variable is allocated before allocating?

```
if (allocated(variable)) deallocate(variable)
allocate(variable)
```

ganganoaa replied:
Tom was right. It will fail if you try to assign an unallocated variable. I tested it in a simple code.

laurenchilutti replied:
@thomas-robinson I wrote a test program to see if we would get an error if `horiz_interp_type_out` is unallocated, and there is no error and the program runs fine.  I found a section of the Fortran 2003 standard that seems to indicate that If a member of horiz_interp_out is unallocated, it will be allocated with the same shape as horiz_interp_in.

You can see the section of the fortran 2003 standard I am referring to [here](https://wg5-fortran.org/N1601-N1650/N1601.pdf#page=155) in Section 7.4.1.3 Interpretation of intrinsic assignments

laurenchilutti replied:
@thomas-robinson Can you take a look at my previous comment on this thread and let me know if you agree with my reasoning?

thomas-robinson replied:
You need to do this for all of the allocatable variables.

thomas-robinson reviewed (changes requested):
I think you need to add `if allocated` statements and change that last pointer to an allocatable unless it's being used somewhere else as a pointer.

ganganoaa commented:
Lauren, the fortran reference verified my assumption and what I have 
tested recently.

Thanks.


On 1/11/23 10:38 AM, laurenchilutti wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In horiz_interp/horiz_interp_type.F90 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1067145573>:
>
> > +    horiz_interp_out%facj            =  horiz_interp_in%facj
> +    horiz_interp_out%ilon            =  horiz_interp_in%ilon
> +    horiz_interp_out%jlat            =  horiz_interp_in%jlat
> +    horiz_interp_out%area_src        =  horiz_interp_in%area_src
> +    horiz_interp_out%area_dst        =  horiz_interp_in%area_dst
> +    horiz_interp_out%wti             =  horiz_interp_in%wti
> +    horiz_interp_out%wtj             =  horiz_interp_in%wtj
> +    horiz_interp_out%i_lon           =  horiz_interp_in%i_lon
> +    horiz_interp_out%j_lat           =  horiz_interp_in%j_lat
> +    horiz_interp_out%src_dist        =  horiz_interp_in%src_dist
>
> @thomas-robinson <https://github.com/thomas-robinson> I wrote a test 
> program to see if we would get an error if |horiz_interp_type_out| is 
> unallocated, and there is no error and the program runs fine. I found 
> a section of the Fortran 2003 standard that seems to indicate that If 
> a member of horiz_interp_out is unallocated, it will be allocated with 
> the same shape as horiz_interp_in.
>
> You can see the section of the fortran 2003 standard I am referring to 
> here <https://wg5-fortran.org/N1601-N1650/N1601.pdf#page=155> in 
> Section 7.4.1.3 Interpretation of intrinsic assignments
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1067145573>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/A43PSQGLLCA6QDXK4XTMXPLWR3HXZANCNFSM6AAAAAATPWEFNQ>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

laurenchilutti commented:
I have updated the `found_neighbors` field to be allocatable.  
Lines 189-191 in hoiz_interp_type.F90 are checking that `horiz_interp_in%I_am_initialized` is set to `.TRUE.` `%I_am_initialized` is set to `.TRUE.` when these fields are allocated, so I will not be adding an `if(allocated(in%field))` before each field assignment in `horiz_interp_type_eq`.
I added `if (allocated(horiz_interp_in%found_neighbors))` before we equate the `found_neighbors` field because this is not allocated in FMS.

rem1776 commented:
@laurenchilutti It looks like the CI glitched out or something on the last commit. I can see that it ran and passed yesterday [here](https://github.com/NOAA-GFDL/FMS/actions?query=branch%3Apointers) but the status never updated on the actual PR. I tried rerunning the latest jobs but it still didn't update.

I've seen something similar before it might've been a github action outage or something at the time you pushed it, so a new commit could fix it. Maybe you can revert/rebase and then re-push the last commit? Although it might be easier to just make some kind of small change and commit that.

laurenchilutti commented:
@rem1776 I just added a minor change to a comment in the latest commit to relaunch the CI.

thomas-robinson reviewed (changes requested):
You need to check that all of the variables are allocated in the assignment routine.  If they aren't allocated, I think that routine will crash.

ganganoaa commented:
I think the check before the assignments,

if(.not.horiz_interp_in%I_am_initialized) then

callmpp_error(FATAL,'horiz_interp_type_eq: horiz_interp_type variable on 
right hand side is unassigned')

endif

will suffice the requirement. If needed, extra safe-guard can be inserted.

Ganga

On 1/20/23 10:15 AM, Tom Robinson wrote:
>
> ***@***.**** requested changes on this pull request.
>
> You need to check that all of the variables are allocated in the 
> assignment routine. If they aren't allocated, I think that routine 
> will crash.
>
> ------------------------------------------------------------------------
>
> In horiz_interp/horiz_interp_type.F90 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#discussion_r1082684259>:
>
> > +    horiz_interp_out%facj            =  horiz_interp_in%facj
> +    horiz_interp_out%ilon            =  horiz_interp_in%ilon
> +    horiz_interp_out%jlat            =  horiz_interp_in%jlat
> +    horiz_interp_out%area_src        =  horiz_interp_in%area_src
> +    horiz_interp_out%area_dst        =  horiz_interp_in%area_dst
> +    horiz_interp_out%wti             =  horiz_interp_in%wti
> +    horiz_interp_out%wtj             =  horiz_interp_in%wtj
> +    horiz_interp_out%i_lon           =  horiz_interp_in%i_lon
> +    horiz_interp_out%j_lat           =  horiz_interp_in%j_lat
> +    horiz_interp_out%src_dist        =  horiz_interp_in%src_dist
>
> You need to do this for all of the allocatable variables.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-GFDL/FMS/pull/1100#pullrequestreview-1263785204>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/A43PSQBS2WZDTNDOWBYX3B3WTKT2FANCNFSM6AAAAAATPWEFNQ>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>
